controllers/machine_controller.go

```diff
@@ -253,32 +272,41 @@ func (r *MachineReconciler) getMachinesInCluster(ctx context.Context, namespace,
 	return machines, nil
 }
 
-// isDeleteReady returns an error if any of Boostrap.ConfigRef or InfrastructureRef referenced objects still exist.
-func (r *MachineReconciler) isDeleteReady(ctx context.Context, m *clusterv1.Machine) error {
-	if m.Spec.Bootstrap.ConfigRef != nil {
-		_, err := external.Get(r.Client, m.Spec.Bootstrap.ConfigRef, m.Namespace)
-		if apierrors.IsNotFound(err) {
-			return nil
+// reconcileDeleteExternal tries to delete external references, returns true if it cannot find any.
+func (r *MachineReconciler) reconcileDeleteExternal(ctx context.Context, m *clusterv1.Machine) (bool, error) {
+	objects := []*unstructured.Unstructured{}
+	references := []*corev1.ObjectReference{
+		m.Spec.Bootstrap.ConfigRef,
+		&m.Spec.InfrastructureRef,
+	}
+
+	// Loop over the references and try to retrieve it with the client.
+	for _, ref := range references {
+		if ref == nil {
+			continue
 		}
-		if err != nil {
-			return errors.Wrapf(err, "failed to get %s %q for Machine %q in namespace %q",
-				path.Join(m.Spec.Bootstrap.ConfigRef.APIVersion, m.Spec.Bootstrap.ConfigRef.Kind),
-				m.Spec.Bootstrap.ConfigRef.Name, m.Name, m.Namespace)
+
+		obj, err := external.Get(r.Client, ref, m.Namespace)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return false, errors.Wrapf(err, "failed to get %s %q for Machine %q in namespace %q",
+				obj.GroupVersionKind(), obj.GetName(), m.Name, m.Namespace)
+		}
+		if obj != nil {
+			objects = append(objects, obj)
 		}
-		return errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 10 * time.Second},
-			"delete is not ready, Bootstrap configuration still exists")
 	}
 
-	if _, err := external.Get(r.Client, &m.Spec.InfrastructureRef, m.Namespace); err != nil && !apierrors.IsNotFound(err) {
-		return errors.Wrapf(err, "failed to get %s %q for Machine %q in namespace %q",
-			path.Join(m.Spec.InfrastructureRef.APIVersion, m.Spec.InfrastructureRef.Kind),
-			m.Spec.InfrastructureRef.Name, m.Name, m.Namespace)
-	} else if err == nil {
-		return errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 10 * time.Second},
-			"delete is not ready, Infrastructure configuration still exists")
+	// Issue a delete request for any object that has been found.
+	for _, obj := range objects {
+		if err := r.Delete(ctx, obj); err != nil {
```

Comment:
This should likely swallow NotFound errors.

Reply:
Shouldn't not found be caught in the `Get` above? Or are you thinking for example if something external between the Get and Delete removes it?

Reply:
It does seem a bit racy to not also check here, but it might be okay because it should pick up on the next reconcile